EHN: Channel keyword for Fan plots

carleyjmartin opened the pull request:
# Scope 

This PR adds a keyword to the plot_fan method which allows the user to plot data from a single channel.
In addition, an exception has been made for when the user chooses a channel but it doesn't exist. 

A warning has also been made which warns that the data chosen has partial records: this is mainly due to the HAN and PYK radars running channels 1 and 2, but channel 2 only has partial records (CPID: 26999 'dummy channel'). But, the warning itself may be useful elsewhere down the line. 

The default will remain at `channel ='all'` which is the same as the RTP plots, and for the majority of cases 'all' will be fine and nothing will change- especially for the channels which are separated into files anyway. 

The main issue is that if you have files with multiple channel data, it may still try to plot 'all' and get confused and plot partial or no data (like the example in the issue), the only thing I can think of to remove this is to set a warning to say 'there's more than one channel of data in here, take care with interpretation and use the channel keyword if needed'. However, this will need to search all the data read in to see which channels are used, and can considerably slow down fan plotting where this isn't really needed 95% of the time. If you think this is a good idea, I can add in this warning, otherwise the user will just have to figure out that the data has more than one channel in the file and that's why it's looking a bit funky. 

**issue:** close #182 

## Approval

**Number of approvals:** 2

## Test

**matplotlib version**: 3.4.2
**Note testers: please indicate what version of matplotlib you are using**
This will return a plot of the channel 1 data:
```python
import matplotlib.pyplot as plt
import pydarn
import datetime as dt

file = "/Users/carley/Documents/data/20141218.0801.00.han.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

time = dt.datetime(2014,12,18,8,24)

fanplot = pydarn.Fan.plot_fan(fitacf_data, scan_index=time,
                    colorbar_label='Velocity [m/s]', lowlat=50, groundscatter=True, channel=1)            
plt.show()
```
<img width="704" alt="Screen Shot 2021-09-09 at 3 56 35 PM" src="https://user-images.githubusercontent.com/60905856/132767970-badbb599-4949-43ba-b482-c88442d18adb.png">

This will return an empty plot and should show the warning that it's a partial record:
```python
import matplotlib.pyplot as plt
import pydarn
import datetime as dt

file = "/Users/carley/Documents/data/20141218.0801.00.han.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

time = dt.datetime(2014,12,18,8,24)

fanplot = pydarn.Fan.plot_fan(fitacf_data, scan_index=time,
                    colorbar_label='Velocity [m/s]', lowlat=50, groundscatter=True, channel=0)            
plt.show()
```
<img width="641" alt="Screen Shot 2021-09-09 at 3 57 06 PM" src="https://user-images.githubusercontent.com/60905856/132768049-bc405460-bdca-4bdf-9346-42d2648397ea.png">

![Screen Shot 2021-09-09 at 3 57 24 PM](https://user-images.githubusercontent.com/60905856/132768055-801913cf-a2ac-46f4-90e4-fd666dd5adc1.png)

And this will give the exception that the channel doesn't exist:
```python
import matplotlib.pyplot as plt
import pydarn
import datetime as dt

file = "/Users/carley/Documents/data/20141218.0801.00.han.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

time = dt.datetime(2014,12,18,8,24)

fanplot = pydarn.Fan.plot_fan(fitacf_data, scan_index=time,
                    colorbar_label='Velocity [m/s]', lowlat=50, groundscatter=True, channel=0)            
plt.show()
```
![Screen Shot 2021-09-09 at 4 05 37 PM](https://user-images.githubusercontent.com/60905856/132768756-af07d4d0-90d4-4a04-986c-98c7a16bdad8.png)


This will return the original plot that gets confused (channel='all'):
```python
import matplotlib.pyplot as plt
import pydarn
import datetime as dt

file = "/Users/carley/Documents/data/20141218.0801.00.han.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

time = dt.datetime(2014,12,18,8,24)

fanplot = pydarn.Fan.plot_fan(fitacf_data, scan_index=time,
                    colorbar_label='Velocity [m/s]', lowlat=50, groundscatter=True)            
plt.show()
```
<img width="610" alt="Screen Shot 2021-09-09 at 3 58 27 PM" src="https://user-images.githubusercontent.com/60905856/132768162-3fa81041-5943-49e8-8499-1318302377b5.png">


Note! that most radars use 0 as the main channel, but a lot use 1 and 2 if using two channels, so you never know.
The HAN data is just an example, but I've tested on SAS a and b files along with single channel files, please test with whatever you have available to get a rounded picture of it working. 
Pytest ran with no failures, not modified at this time.

mts299 commented:
@carleyjmartin Code looks good, can you add a pytest for this parameter change?

mts299 commented on pydarn/exceptions/plot_exceptions.py:
```diff
@@ -175,3 +179,16 @@ def __init__(self, parameter: str, gate_num: int, max_range_gate: int):
+    def __init__(self, channel: int):
+        self.channel = channel
+        self.message = "There is no data for channel {channel}"\
+                       .format(channel=self.channel)
```
Similar to my no data warning maybe we could suggest a channel that data has. This would help with say "twofsound" when we have 1 and 2 and when we have stero mode with 0 and 1

carleyjmartin replied:
I've added the option to give an option in the exception, but the implementation of the exception in fan.py is a bit rudimentary at the moment, assuming I will probably change how it works when I sort out the other issue below. 


(I also know why it's doing that other issue, but need to have a think about how to go around it so it preserves the fix for the issue with the HAN data too)

mts299 commented:
So I got some twofsound data and the channels  change per scan but when I do a `scan_index` and switch between 1 and 2 it comes up with data and not one with "no data" error
``` python
import pydarn
from datetime import datetime
import matplotlib.pyplot as plt 

rkn_file = 'data/20180101.0000.01.rkn.fitacf'
rkn_data = pydarn.SuperDARNRead().read_dmap(rkn_file)
ydarn.Fan.plot_fan(rkn_data, scan_index=3,
                    colorbar=False, line_color='blue',
                    radar_label=True, channel=1)
```

`channel=1`
![channel1](https://user-images.githubusercontent.com/6520530/132884626-302c9605-ccb2-4f12-bfd8-4718389fb8fd.png)

`channel=2`
![channel2](https://user-images.githubusercontent.com/6520530/132884699-40290fdd-dba7-4e59-b9ea-f4003d2b18bb.png)

mts299 reviewed (changes requested):
Look into the twofsound scan index issue

carleyjmartin commented:
So here's the dilemma with this: at the moment if a channel is given in the function call for a fan, then the code will remove any data that isn't in that channel and then proceed with the rest of the fan plotting. This works well for HAN and PYK, as their scans are interleaved with each other i.e:
scan flag:1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0...
channel:  0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1...
So the channel switches between each data point in the file, kind of blurring what it means to have a '1st' scan, where is it the full cycle above that is one scan or two scans? As you can see, just removing one channel is fine for this issue as we're not too set on numbering of the scans as they start/end at roughly the same time, and selecting channel 1 or 0 will give you a nice 16 beams scan. 

HOWEVER, we then have the opposite issue with RKN twofsound data, where the scan and channels look like this:
scan flag: 1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0...
channel:   1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2...
So using channel='all' doesn't break here, and we can just use that, but then if you select only channel 1 and rather than inputting a time, input a scan index N, it will give you the Nth scan of ONLY that channel data, not the entire data set. But inputting a time will either plot it correctly or tell you there's no data as it's the other channel operating at that minute, which is as expected.
The Nth scan index issue can be fixed by just including a channel check later in the if statement (same as RTP) and removing the current solution. But, then we would be back to it not working for the HAN and PYK channel/scan-flag pattern. I'm not sure if it's possible to include both of these solutions without some pretty horrific if loops and/or convoluted pattern detection that will considerably slow the plotting down - we can't use cpid detection as the cpid can be different for each channel.

I might suggest that we can specify that the option to 'choose a channel' is for scans that are operating on two channels simultaneously rather than one scan after the other, as the one after the other scans can easily be accessed using the channel='all' option, whereas the simultaneous scans can't and require some filtering of data to plot. I'll keep having a poke at it, and if there are any suggestions let me know but as of now I can't think of a non-convoluted way of keeping both of these solutions in the code without further confusion or issues.

Anyway, I hope that makes sense. TLDR: combination of different scan_flag and channel patterns mean there's no single all encompassing solution to plotting a specific channel/scan for fan plots.

mts299 commented:
@carleyjmartin still reading the novel response, but shouldn't we look for the scan first then plot if the channel is correct? Would this not fix the problem. 

I worry if we remove the channel data then find the scan then we are not looking accurately at scan 3 anymore

carleyjmartin commented:
@mts299  The main issue I can't think around is how to pick out the scans from the sequentially scanning data? Is the 'first scan' the entire '1,1,0,0,0,0,0,0,0,0,0,0,..' or is that technically two scans? How can we tell the code to sometimes ignore that second 1? Which I guess is the original issue of scan_flag being used to build the scans, when in this case the scan_flag is only signalling the new scan for a single channel, and then the next record goes back to a previous scan. 

I think I have an idea anyway, will have another crack tomorrow.

mts299 commented:
Okay message me on slack for more idea thinking.

carleyjmartin commented:
Added channel option to fan plot read the docs, also added a note under the options to discuss that in some cases you might NEED to use a channel integer, not the default 'all'. And that the scan_index will be counting scans of the specific channel if a channel is given. This is changed in the docs and in the docstring in fan.py. Slightly out of scope, but very minor change: also added the channel option the range-time docs, I think there was confusion long ago about RTP taking a channel option, and it was just accidentally missed off in the docs.
